pkg/registration/offline_certificate.go

```diff
@@ -0,0 +1,108 @@
+package registration
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+)
+
+// The information extracted from an offline registration certificate
+// To further inspect payload and signature check [IsValid] and [ExtractPayload].
+type OfflineCertificate struct {
+	Version          string `json:"version"`
+	Cipher           string `json:"cipher"`
+	Hash             string `json:"hash"`
+	EncodedPayload   string `json:"payload"`
+	EncodedSignature string `json:"signature"`
+}
+
+// The information supplied and validated by the payload. The [RegcodeHash] can be used
+// to determine if the certificate hold has the knowledge over the registration code.
+//
+// Information holds arbitrary information about the registered system and might or might
+// not have the relevant information. Make sure to check if the key exists and prepare for
+// type casting if necessary.
+type OfflinePayload struct {
+	Login        string         `json:"login"`
+	Password     string         `json:"password"`
+	Subscription Subscription   `json:"subscription"`
+	RegcodeHash  string         `json:"regcode_sha256"`
+	Information  map[string]any `json:"information"`
+}
+
+// Reads an offline registration certificate from a read object
+//
+// An error indicates the certificate was probably malformed
+func OfflineCertificateFrom(reader io.Reader) (*OfflineCertificate, error) {
+	certificate := OfflineCertificate{}
```

Comment:
Declare this is a pointer `certificate := &OfflineCertificate{}` and then you won't have to de-reference it in `json.Unmarshal` or when returning it.